src/generator.cs

```diff
@@ -1117,6 +1127,176 @@ static string FindProtocolInterface (Type type, MemberInfo pi)
+			temp = string.Format ("new NSNumber ({1}{0});", isNullable ? ".Value" : string.Empty, pi != null ? pi.Name.GetSafeParamName () : "value");
+
+		else if (originalType == typeof (NSValue)) {
+			var typeStr = string.Empty;
+			if (!NSValueCreateMap.TryGetValue (retType, out typeStr)) {
+				// HACK: These are problematic for X.M due to we do not ship System.Drawing for Full profile
+				if (retType.Name == "RectangleF" || retType.Name == "SizeF" || retType.Name == "PointF")
+					typeStr = retType.Name;
+				else
+					throw new BindingException (1049, true, GetBindAsExceptionString ("box", retType.Name, originalType.Name, "container", minfo?.mi?.Name ?? pi?.Name));
+			}
+			temp = string.Format ("NSValue.From{0} ({2}{1});", typeStr, isNullable ? ".Value" : string.Empty, pi != null ? pi.Name.GetSafeParamName () : "value");
+		} else if (originalType == typeof (NSString) && IsSmartEnum (retType)) {
+			temp = string.Format ("{1}{0}.GetConstant ();", isNullable ? ".Value" : string.Empty, pi != null ? pi.Name.GetSafeParamName () : "value");
+		} else if (originalType.IsArray) {
+			var arrType = originalType.GetElementType ();
+			var arrRetType = Nullable.GetUnderlyingType (retType.GetElementType ()) ?? retType.GetElementType ();
+			var valueConverter = string.Empty;
+
+			if (arrType == typeof (NSString))
+				valueConverter = "o{0}.GetConstant (), {1});";
+				//temp = string.Format ("NSArray.FromNSObjects (o => o{0}.GetConstant (), {1});", isNullable ? ".Value" : string.Empty, pi != null ? pi.Name.GetSafeParamName () : "value");
+			else if (arrType == typeof (NSNumber))
+				valueConverter = string.Format ("new NSNumber ({0}o{{0}}), {{1}});", arrRetType.IsEnum ? "(int)" : string.Empty);
+				//temp = string.Format ("NSArray.FromNSObjects (o => new NSNumber ({2}o{0}), {1});", isNullable ? ".Value" : string.Empty, pi != null ? pi.Name.GetSafeParamName () : "value", arrRetType.IsEnum ? "(int)" : string.Empty);
+			else if (arrType == typeof (NSValue)) {
+				var typeStr = string.Empty;
+				if (!NSValueCreateMap.TryGetValue (arrRetType, out typeStr)) {
+					if (arrRetType.Name == "RectangleF" || arrRetType.Name == "SizeF" || arrRetType.Name == "PointF")
+						typeStr = retType.Name;
+					else
+						throw new BindingException (1049, true, GetBindAsExceptionString ("box", arrRetType.Name, originalType.Name, "array", minfo?.mi?.Name ?? pi?.Name));
+				}
+				valueConverter = string.Format ("NSValue.From{0} (o{{0}}), {{1}});", typeStr);
+				//temp = string.Format ("NSArray.FromNSObjects (o => NSValue.From{2} (o{0}), {1});", isNullable ? ".Value" : string.Empty, pi != null ? pi.Name.GetSafeParamName () : "value", typeStr);
+			} else
+				throw new BindingException (1048, true, "Unsupported type {0} decorated with [BindAs]", retType.Name);
+			temp = string.Format ($"NSArray.FromNSObjects (o => {valueConverter}", isNullable ? ".Value" : string.Empty, pi != null ? pi.Name.GetSafeParamName () : "value");
```

Comment:
There are a few nested string.Formats here, which made this section a bit hard to read.

Maybe something like this would make it easier:

    var denullify = isNullable ? ".Value" : string.Empty;
    var value = pi != null ? pi.Name.GetSafeParamName () : "value";
    if (arrType == typeof (NSString)) {
        valueConverter = $"o{denullify}.GetConstant (), {value}";
    } else if (arrType == typeof (NSNumber)) {
        var cast = arrRetType.IsEnum ? "(int) " : string.Empty;
        valueConverter = $"new NSNumber ({cast}o{denullify}, {value})";
    } else if (arrType == typeof (NSNumber)) {
        var typeStr = ...
        valueConverter = $"NSValue.From{typeStr} (o{denullify}, {value})";
    }
    temp = $"NSArray.FromNSObjects (o => {valueConverter}");